### ISSUE TITLE
ngf-select TypeError: Illegal invocation

### ISSUE THREAD

**Issue by danielbonnell** [danialfarid/ng-file-upload#1537]
I'm running on 12.0.4 (see note at end) with Angular 1.5.5. I have the following code in my view:

    <div class="btn btn-default" ngf-select ngf-multiple="false" ngf-max-size="10MB" ng-model="editSpotlight.logo" name="logo" ngf-pattern="image/*" accept="image/*" ngf-max-size="10MB" required>
        Select Logo
        <i class="fa fa-check" ng-show="editSpotlight.logo"></i>
    </div>

When I select the file, I get the following error in the console:

    angular.self-208d6d1….js?body=1:13551 TypeError: Illegal invocation
        at equals (angular.self-208d6d1….js?body=1:1114)
        at equals (angular.self-208d6d1….js?body=1:1114)
        at equals (angular.self-208d6d1….js?body=1:1098)
        at equals (angular.self-208d6d1….js?body=1:1098)
        at Scope.$digest (angular.self-208d6d1….js?body=1:17067)
        at Scope.$apply (angular.self-208d6d1….js?body=1:17338)
        at $$debounceViewValueCommit (angular.self-208d6d1….js?body=1:26913)
        at $setViewValue (angular.self-208d6d1….js?body=1:26885)
        at update (ng-file-upload.self-3e79cca….js?body=1:530)
        at ng-file-upload.self-3e79cca….js?body=1:612(anonymous function) @ angular.self-208d6d1….js?body=1:13551(anonymous function) @ angular.self-208d6d1….js?body=1:10226Scope.$digest @ angular.self-208d6d1….js?body=1:17092Scope.$apply @ angular.self-208d6d1….js?body=1:17338$$debounceViewValueCommit @ angular.self-208d6d1….js?body=1:26913$setViewValue @ angular.self-208d6d1….js?body=1:26885update @ ng-file-upload.self-3e79cca….js?body=1:530(anonymous function) @ ng-file-upload.self-3e79cca….js?body=1:612(anonymous function) @ angular.self-208d6d1….js?body=1:19158completeOutstandingRequest @ angular.self-208d6d1….js?body=1:5870(anonymous function) @ angular.self-208d6d1….js?body=1:6146
    2016-05-18 10:19:31.927 angular.self-208d6d1….js?body=1:13551 TypeError: Illegal invocation
        at equals (angular.self-208d6d1….js?body=1:1114)
        at equals (angular.self-208d6d1….js?body=1:1114)
        at equals (angular.self-208d6d1….js?body=1:1098)
        at equals (angular.self-208d6d1….js?body=1:1098)
        at Scope.$digest (angular.self-208d6d1….js?body=1:17067)
        at Scope.$apply (angular.self-208d6d1….js?body=1:17338)
        at angular.self-208d6d1….js?body=1:19167
        at completeOutstandingRequest (angular.self-208d6d1….js?body=1:5870)
        at angular.self-208d6d1….js?body=1:6146(anonymous function) @ angular.self-208d6d1….js?body=1:13551(anonymous function) @ angular.self-208d6d1….js?body=1:10226Scope.$digest @ angular.self-208d6d1….js?body=1:17092Scope.$apply @ angular.self-208d6d1….js?body=1:17338(anonymous function) @ angular.self-208d6d1….js?body=1:19167completeOutstandingRequest @ angular.self-208d6d1….js?body=1:5870(anonymous function) @ angular.self-208d6d1….js?body=1:6146

I just migrated from Angular 1.2.18 to 1.5.5 yesterday. This wasn't happening before that. I was also previously using v11.0.0 before/after that migration, which is when I first noticed this issue. I upgraded to v12.0.4 to see if that would have any effect, but the issue persisted.

**Comment by danialfarid** [comment]
Could you create a reproducible jsfiddle?

**Comment by danielbonnell** [comment]
I've been trying unsuccessfully to reproduce it in a fiddle. [This](https://jsfiddle.net/na0f5uL7/1) is as close as I've gotten to mirroring my setup on a fiddle, but it doesn't break when I select an attachment. As far as I can tell, it happens precisely when I select a file, but only on one of my uploaders. I have several uploaders in the app, but they are all virtually identical (same form structure, same controller structure). It also happens very occasionally if I reload a view within the app that contains an uploader.

**Comment by danielbonnell** [comment]
For what it's worth, I tried following the stack trace through the source (which I'm no terribly familiar with) and was eventually able to determine that this function is the source of the problem. Starting on line 610 of `ng-file-upload.js`:

    $timeout(function () {
        update(keep ? prevValidFiles.concat(valids) : valids,
        keep ? prevInvalidFiles.concat(invalids) : invalids,
        files, dupFiles, isSingleModel);
    }, options && options.debounce ? options.debounce.change || options.debounce : 0);

When I inspect with `debugger`, I can see that calling `update()` trips the error. Additionally, `options` appears to be `undefined`.

**Comment by danielbonnell** [comment]
I think I figured out the source of the problem. It's related to [#776](https://github.com/danialfarid/ng-file-upload/issues/776). I had a `$watch` set on an array of elements returned from a data call service. I then assigned that array to a variable in my controller, and assigned an element of that array to my edit form variable. Apparently though, the `$watch` in the data service still covered the form variable in my controller. I removed the `$watch` from the data service and the problem went away.
